manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+// ErrPortSpaceExhausted
+func IsErrPortSpaceExhausted(e error) bool {
+	_, ok := e.(ErrPortSpaceExhausted)
+	return ok
+}
+
+// ErrInvalidEndpoint is returned if the endpoint passed to Allocate or
+// Deallocate is somehow not valid. The error message will contain the problem.
+// It is a catchall for any errors not otherwise covered.
+type ErrInvalidEndpoint struct {
+	problem string
+}
+
+// Error returns a formatted message explaining what the problem is
+func (e ErrInvalidEndpoint) Error() string {
+	return e.problem
+}
+
+// IsErrInvalidEndpoint returns true if the provided error is of type
+// ErrInvalidEndpoint
+func IsErrInvalidEndpoint(e error) bool {
+	_, ok := e.(ErrInvalidEndpoint)
+	return ok
+}
+
+// Allocator is an allocator component that manages the state of which
+// ports (and which protocols) are in use in the the cluster.
+//
+// The Allocator totally owns the Endpoint.Ports field. It provides a slice
+// of PortConfigs which should be written directly without modification to the
+// the Endpoint.Ports field. Nothing should be modified about Endpoint.Ports
+// outside of the port allocator. If this rule is followed, the the port
+// allocator doesn't have to check for cases like two services having the same
+// port allocated; we created all of the fields, and we can know they are
+// consistent.
+type Allocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the Allocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// Proposal is the return value of Allocate and DeallocateEndpoint,
+// which contains the data needed to affirmatively alter the state of the
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	//
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *Allocator
+	ports      []*api.PortConfig
```

Comment:
Is `ports` and the getter function `*proposal.Ports()` being used for anything? It seems they are not needed, maybe some deprecated code?

Reply:
Yes. it's the most important part, used by the caller. The alternative to `Ports()` would be to return `(Proposal, []*api.PortConfig, error)` from `Allocate`. That might be cleaner. WDYT?

Reply:
Oh, I see, you need to modify the given []*api.PortConfig (e.g. dynamic assigned ports) but you can not do it over the given pointer as the whole two-phase model would get broken.

Between the two proposed options, I would suggest sticking to the first because all the information is contained in the proposal, you do not have to keep two variables to keep information that does not have any meaning by its own.